ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+      }
+    }
+    // last offset to compact should be outside the range of the journal
+    Offset lastSegmentEndOffset = new Offset(lastSegment.getName(), lastSegment.getEndOffset());
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
+      throw new IllegalArgumentException("Some of the offsets provided for compaction are within the jounral");
+    }
+  }
+
+  /**
+   * Copies valid data from the source log segments into the swap spaces.
+   * <p/>
+   * Splits the compaction into two cycles if there aren't enough swap spaces and completes the copy for the current
+   * cycle.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void copy() throws InterruptedException, IOException, StoreException {
+    setupState();
+    List<String> logSegmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+    Offset startOffsetOfLastIndexSegmentForDelete = getStartOffsetOfLastIndexSegmentForDelete();
+    FileSpan duplicateSearchSpan = null;
+    if (compactionLog.getCurrentIdx() > 0) {
+      // only the very first log segment in the cycle can contain duplicate records.
+      LogSegment logSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(logSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the previous log segment.
+        Offset startOffset = new Offset(prevSegment.getName(), prevSegment.getStartOffset());
+        Offset endOffset = new Offset(prevSegment.getName(), prevSegment.getEndOffset());
+        duplicateSearchSpan = new FileSpan(startOffset, endOffset);
+      }
+    }
+
+    for (String logSegmentName : logSegmentsUnderCompaction) {
+      LogSegment srcLogSegment = srcLog.getSegment(logSegmentName);
+      Offset logSegmentEndOffset = new Offset(srcLogSegment.getName(), srcLogSegment.getEndOffset());
+      if (needsCopying(logSegmentEndOffset) && !copyLogSegmentData(srcLogSegment, duplicateSearchSpan,
+          startOffsetOfLastIndexSegmentForDelete)) {
+        if (isActive) {
+          compactionLog.splitCurrentCycle(logSegmentName);
+        }
+        // else shutdown is in progress.
+        break;
+      }
+      // only the very first log segment in the cycle can contain duplicate records.
+      duplicateSearchSpan = null;
+    }
+
+    numSwapsUsed = tgtIndex.getLogSegmentCount();
+    tgtIndex.close();
+    tgtLog.close();
+    // persist the bloom of the "latest" index segment if it exists
+    if (numSwapsUsed > 0) {
+      tgtIndex.getIndexSegments().lastEntry().getValue().map(true);
+    } else {
+      // there were no valid entries copied, return any temp segments back to the pool
```

Comment:
This can happen if there is no valid data at all in the segments being compacted.

Cleanup is for cleaning up the source index and log. This special case is for the target log.